design/secure-communication-design.md

```diff
@@ -0,0 +1,88 @@
+# Secure Communication between Contour and Envoy
+
+_Status_: Draft, in review.
+
+This document outlines how to secure the communication between Contour
+and Envoy containers.
+
+## Goals
+
+- Achieve secure communication between Contour and Envoy
+- Authenticate Envoy connecting to Contour
+- Rotating and invalidating the credentials
+
+## Non-goals
+
+- Support for dynamic credentials
+- Support for use of Secret Discovery Service for credential management
+
+## Background
+
+Currently, Contour and Envoy containers are co-located within a pod.
+Current security model is relying on the fact that Contour exposes the management
+server on loopback interface which prevents any unauthorized access from outside
+the pod.
+
+With the need to move away from the [current deployment pattern of co-locating
+Contour and Envoy][0], it is necessary to secure the communication channel between
+the containers.
+
+## High level design
+
+- Use mutual TLS to secure communication channel between Contour and Envoy
+- Bootstrap phase of Contour will generate necessary TLS certificates
+
+## Detailed design
+
+### Default options for secure communication
+
+By default, mutual TLS is enabled to secure the communication between Contour and Envoy
+containers. The default `contour bootstrap` will generate the bootstrap configuration that
+enables secure communication between the containers. If `contour bootstrap` is passed with
+the parameter `--disable-tls`, the bootstrap configuration generated will make use of insecure
+communication between the containers.
+
+### Bootstrapping process
+
+This bootstrap phase is independent of `contour bootstrap` and uses `contour bootstrap-tls`.
+`contour bootstrap-tls` is run as an init container to bootstrap the TLS configuration.
+The bootstrap phase will create three pairs of the TLS certificates. The details about
+each pair of certificates are mentioned below:
+1. The first pair is used by Contour to serve HTTPS.
+2. The second pair is used by Envoy to authenticate with Contour.
+3. The last pair serves as the CA certificate which signs the above certificates.
+
+Each of these TLS certificates will be stored as Kubernetes secrets in namespace `heptio-contour`. The
```

Comment:
There might be a chicken & egg problem here. If we utilize Kubernetes secrets, the pod won't start until the secret exists. We may need to do something like how Sonobuoy does it by having a `contour gen` command which would give you the secrets that would be applied to cluster.

Reply:
As far as I know, as the init containers brings up the Kubernetes secrets, the pod will be able to start.

Reply:
My initial thought was that in the contour yaml, we'd reference a secret. This would allow any instance of Contour to block from starting until the secret exists. If this is the case, I think we need some external process to generate the secret which Contour blocks on starting until that secret is available. The Contour deployment resource needs to reference a secret, it won't start until that secret is available. 

If Contour doesn't mount the secret as a volume, then I'm still not sure how multiple instances of Contour can coordinate together and not overwrite each other. It would work if you bring up a single instance of Contour at a time, but we can't guarantee folks will do that.

Reply:
It would take a larger discussion, but we could look at pulling in some of the kubebuilder bits to simply the CRD work, as well as give us leader election. I haven't used before though so can't speak to how well it works: https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/manager/manager.go#L104-L114

Reply:
There may be a way we can do leader election with a configmap lock (https://github.com/kubernetes/client-go/blob/master/examples/leader-election/main.go). 

Looks like that example is broken, I'm working to fix it now.

Reply:
Like I had mentioned before, we can have the init container to create the necessary secrets for us. I agree that we might want this to performed by one of the pods and not all of the pods in split deployment scenarios. This can definitely make use of the leader election provided by the Kubernetes client-go. 

For reference, for how init containers can create secrets: https://github.com/kelseyhightower/certificate-init-container.